lib/meta-box-partial-page.php

```diff
@@ -550,3 +554,115 @@ function gutenberg_is_meta_box_empty( $meta_boxes, $context, $post_type ) {
 }
 
 add_filter( 'filter_gutenberg_meta_boxes', 'gutenberg_filter_meta_boxes' );
+
+/**
+ * Go through the global metaboxes, and override the render callback, so we can trigger our warning if needed.
+ *
+ * @since 1.8.0
+ */
+function gutenberg_intercept_meta_box_render() {
+	global $wp_meta_boxes;
+
+	foreach ( $wp_meta_boxes as $post_type => $contexts ) {
+		foreach ( $contexts as $context => $priorities ) {
+			foreach ( $priorities as $priority => $boxes ) {
+				foreach ( $boxes as $id => $box ) {
+					if ( ! is_array( $wp_meta_boxes[ $post_type ][ $context ][ $priority ][ $id ]['args'] ) ) {
+						$wp_meta_boxes[ $post_type ][ $context ][ $priority ][ $id ]['args'] = array();
+					}
+					if ( ! isset( $wp_meta_boxes[ $post_type ][ $context ][ $priority ][ $id ]['args']['__original_callback'] ) ) {
+						$wp_meta_boxes[ $post_type ][ $context ][ $priority ][ $id ]['args']['__original_callback'] = $box['callback'];
+						$wp_meta_boxes[ $post_type ][ $context ][ $priority ][ $id ]['callback']                    = 'gutenberg_override_meta_box_callback';
+					}
+				}
+			}
+		}
+	}
+}
+add_action( 'submitpost_box', 'gutenberg_intercept_meta_box_render' );
+add_action( 'submitpage_box', 'gutenberg_intercept_meta_box_render' );
+add_action( 'edit_page_form', 'gutenberg_intercept_meta_box_render' );
+add_action( 'edit_form_advanced', 'gutenberg_intercept_meta_box_render' );
+
+/**
+ * Check if this metabox only exists for back compat purposes, show a warning if it doesn't.
+ *
+ * @since 1.8.0
+ *
+ * @param mixed $object The object being operated on, on this screen.
+ * @param array $box The current meta box definition.
+ */
+function gutenberg_override_meta_box_callback( $object, $box ) {
+	$callback = $box['args']['__original_callback'];
+	unset( $box['args']['__original_callback'] );
+
+	$block_compatible = true;
+	if ( isset( $box['args']['__block_editor_compatible_meta_box'] ) ) {
+		$block_compatible = (bool) $box['args']['__block_editor_compatible_meta_box'];
+		unset( $box['args']['__block_editor_compatible_meta_box'] );
+	}
+
+	if ( isset( $box['args']['__back_compat_meta_box'] ) ) {
+		$block_compatible |= (bool) $box['args']['__back_compat_meta_box'];
+		unset( $box['args']['__back_compat_meta_box'] );
+	}
+
+	if ( ! $block_compatible ) {
+		gutenberg_show_meta_box_warning( $callback );
+	}
+
+	call_user_func( $callback, $object, $box );
+}
+
+/**
+ * Display a warning in the metabox that the current plugin is causing the fallback to the old editor.
+ *
+ * @since 1.8.0
+ *
+ * @param callable $callback The function that a plugin has defined to render a meta box.
+ */
+function gutenberg_show_meta_box_warning( $callback ) {
+	// Only show the warning when WP_DEBUG is enabled.
+	if ( ! WP_DEBUG ) {
```

Comment:
I wonder if we should not show it even in "prod" while we're on the plugin phase and introduce this change later. To encourage people to check their metaboxes?

Reply:
I did this way while we're still working out the path for plugin devs to move their meta boxes to blocks - it doesn't seem fair to show a warning that the dev can't do anything to fix.

Once there are well documented upgrade paths, we can take away this check. Once it comes time to merge, we can evaluate how we want to merge it. It might even be valuable to have a "plugin x caused this fall back" message in prod, so users know what happened, and a "here are links to documentation and tutorials to upgrade your meta boxes" when `WP_DEBUG` is set, so plugin devs have the information they need.